### PR TITLE
non flash attention: speedup by avoiding ddp broadcasts of causal mask

### PR DESCRIPTION
**Pull request by YassineYousfi**
In the manual implementation of causal self-attention, the causal mask is registered as a buffer, which causes DDP to broadcast it at every step. Excluding it from being broadcasted gives some extra speedup (up to 8% in a 8x3090 setup with fake data, probably more for multinode)

In fact, the causal masks are the only buffers in this model, so you can also set `broadcast_buffers=False` in train.py:L201 but that could potentially backfire if other buffers that need syncing are introduced in the future.

### CODE REVIEW

**Comment by vgoklani** [comment]
that's a really clever observation, thanks for sharing!

did not know about this: "self._ddp_params_and_buffers_to_ignore"

https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L565

:)
